### ISSUE TITLE
Build physics collider engine

### ISSUE THREAD

**Issue by jacobgardner** [jacobgardner/hack-and-slash#5]
My idea for this is to make all movement work by placing a velocity on entities, which show the direction you want the entity to move.  The physics 'engine' will resolve the vector and determine if the movement was valid or not, then updating the movements accordingly.  Might be the wrong way to approach this.

**Comment by jrhii** [comment]
So to clarify, the physics engine will be fed all desired velocities and positions of all relevant entities and then update both velocity and position (or maybe just position is needed? does the engine handle gravity?) according to whether or not they meet (and if we have non-colliding entities, handle those?)

If this handles gravity, then moving entities only need to have simple vector and speed controls, I assume>

I'm thinking along the lines of the entity wants to move along an axis at their given speed (a constant each one might hold) whatever provides velocities to this engine would have already decided if an entity has the capability to move (e.g. if it's airborne, it can't give an upward (jumping) velocity unless it's object properties allow for double-jumping or flying) and the engine decides if nothing is blocking the entity (e.g. other entities, or in case of gravity the ground)

It could also potentially update each entity with an on_ground property if it finds no updates with the y axis speed.

Am I on the right track here?

**Comment by jacobgardner** [comment]
So I looked at a real API for physics.  Looks like what they do is:
- They have a `PhysicsBody` class which can be rectangle/circle/custom.  This contains the object's position, velocity and acceleration (excluding gravity).  So we would not actually modify the entity, but this `physicsBody` object in the entity. 
- The 'engine' bit gets fed in all of these `physicsBody` objects before running and then you hit `resolve()`
- Resolve goes through and updates all the positions and velocities appropriately.  
- Then we go through all our entities, and hit render()
- When drawing our entities, we look at that physicsBody object for the appropriate positions and use that as a reference for drawing. 
- Also the physics engine typically will have options for gravity/etc. So for jumping up, you would just set the velocity vector to a positive y and the physics engine would take care of everything until you hit the ground again.

Did that answer all your questions?

**Comment by jacobgardner** [comment]
Yeah, things like when to jump, and external forces will be given to the `physicsBody` objects outside of the physics part of the engine.  It's really just the physics engine's job to resolve the positions of everything given their shape, velocity, position, and acceleration.

**Comment by jacobgardner** [comment]
It might also be useful for the physics engine to not block certain `physicsBody` interactions.  But still be able to report collisions.  

For example, maybe we want the player to be able to move through enemies, but we'll rely on the physics engine to tell us they did intersect for purposes of applying damage to the player or the enemy.

**Comment by jrhii** [comment]
OK, that's similar to how i envisioned it.  clears some things up, yes

**Comment by jrhii** [comment]
so
for the physics
Each relevant entity contains their physicsBody object
and at the beginning of a scene, all entity's are loaded into the engine
which would only need to be loaded when they are created for a given scene.  after that the objects in the entity and engine will be referencing the same memory.  then the engine should only alter the physicsBody 's upon a resolve() or some such command
I can imagine storing simple simple geometric "Hit boxes" (whatever it is called for detecting an actual collision) easily and efficiently enough.  storing more complex shapes, I can't think of a method off the top of my head other then it being a composite of simple shapes
if we need to go that route
e.g.
Complex: [ { Type: "Circle", coordinates : {r, theta }}, {Type: Square, {r, theta}}]
and type would be like, a simple shape object
coordinates is its location from the center of the body
But for sake of time, I am advocating for simple shapes unless things are looking optimistic or we really need geometrics
https://www.npmjs.com/package/polygon
polygon
utility for working with polygons (arrays of vec2s)
npmjs.com

Jacob
Sorry was driving
i eat at a place called schneithorsts and then watch a foreign film
in the meantime you can post any thought on the matter
TUE 9:17PM

Jacob
I can't imagine us having complex shapes
I think the hardest part of this will be efficiently looping through all the entities
My initial thought is bucketing the entities based on their position
So essentially imagine a grid over the entire level
make the grid size configurable
And so for each grid that has at least one physicsBody, make an array and put those in there.
Then when we put a velocity on a physicsBody, we 'activate' that cell and the surrounding cells
and we only do resolves on those active cells
That polygon library looks cool
Curious how you're going to deal with collisions
TUE 10:49PM
Simply, at first.  No weird balance or center of gravity for vertical, if that's what you mean

Jacob
Well there shouldn't be balance in a 2d platformer
Like
Let's say a square character falls on a triangle
Without balance of w.e.  it just stay where it hits
Not tilting or falling down a slope

Jacob
Let's make triangles = spikes = death

**Comment by jacobgardner** [comment]
Just FYI Jim, I'm planning on using your physics engine to determine what to draw to the screen.  I'll make a rectangle `physicsBody` that represents the current screen coordinates in the world, then find all the `physicsBodies` it collides with and draw those.  That way if we do optimization on the physics engine side, it automatically carries forward to the renderer code.

**Comment by jrhii** [comment]
said `rectangle` would be the "camera" then?

**Comment by jacobgardner** [comment]
Yeah

**Comment by jrhii** [comment]
#basic collider implementation goal:  

##Use simple axis aligned bounding box from the polygon object to detect collisions.

##do a broad search with a Quadtree....once I finish reading about what exactly that is
